torch/testing/_internal/common_device_type.py

```diff
@@ -312,6 +320,17 @@ def rel_tol(self, prec):
     def get_primary_device(cls):
         return cls.device_type
 
+    @classmethod
+    def _init_and_get_primary_device(cls):
+        try:
+            return cls.get_primary_device()
+        except Exception:
+            # For CUDATestBase, XLATestBase, and possibly others, the primary device won't be available
+            # until setUpClass() sets it. Call that manually here if needed.
+            if hasattr(cls, 'setUpClass'):
+                cls.setUpClass()
+            return cls.get_primary_device()
```

Comment:
This is a bit messy. The device specific test class for CUDA, XLA, and maybe others (?) define `setupUpClass()` method that populates the primary device. In the past, this has been run by `unittest` just before the first test invocation. Unfortunately, that's too late if we want the per-param decoration mechanism to have access to it during test generation time.

The hacky approach here calls `setUpClass()` if trying to get the primary device fails. Open to better suggestions! Note that I have been operating under the assumption that there may be other device-specific test classes out there we don't know about or can't easily change. If we aren't restricted by this, other options involving changing `CUDATestBase` / `XLATestBase` open up.

Reply:
I also can't immediately think of a better way to do this; I found the comment very helpful